### ISSUE TITLE
ERR: validate boolean indexing with iloc

### ISSUE THREAD

**Issue by jreback** [pandas-dev/pandas#8976]
xref #8956

This should raise a `ValueError`.

```
In [18]: s = Series(np.arange(5,1,-1))

In [19]: s
Out[19]: 
0    5
1    4
2    3
3    2
dtype: int64

In [20]: s.iloc[[True,False,True]]
Out[20]: 
0    5
2    3
dtype: int64
```

**Comment by jorisvandenbossche** [comment]
In the meantime, numpy has changed this behaviour (already deprecated in 1.10, errors in 1.13), so we should follow I think. This should raise:

```
In [1]: np.__version__
Out[1]: '1.13.3'

In [2]: pd.__version__
Out[2]: '0.22.0.dev0+202.g492040b'

In [3]: a = np.arange(5)

In [4]: a[np.array([True, False, True])]
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-4-7375c598447e> in <module>()
----> 1 a[np.array([True, False, True])]

IndexError: boolean index did not match indexed array along dimension 0; dimension is 5 but corresponding boolean dimension is 3

In [5]: s = pd.Series(a)

In [6]: s.iloc[np.array([True, False, True])]
Out[6]: 
0    0
2    2
dtype: int64

```

**Comment by jreback** [comment]
yep agreed. I think they wrapped around the booleans before.

**Comment by jbrockmendel** [comment]
We now raise `IndexError: Item wrong length 3 instead of 4.` on this which looks about right

**Comment by jbrockmendel** [comment]
Closed by #26911
